File: data.structures/haskell/Demos/Queue/Josephus.hs
-------------------------------------------------------------------------------
-- Solution to Josephus' problem using a queue
--
-- Data Structures. Grado en Informática. UMA.
-- Pepe Gallardo, 2012
-------------------------------------------------------------------------------

module Demos.Queue.Josephus where

import DataStructures.Queue.TwoListsQueue

mkQueue :: Int -> Queue Int
mkQueue n = foldr enqueue empty [n,n-1..1]

skip :: Int -> Queue a -> Queue a
skip n q
   | n==0 || isEmpty q = q
   | n>0               = skip (n-1) (enqueue x q')
   where
     x = first q
     q' = dequeue q

josephus :: Int -> Int -> [Int]
josephus n m = execute $ mkQueue n
 where
  execute q
   | isEmpty q = []
   | otherwise = x : execute (skip m q')
   where
     x = first q
     q' = dequeue q

File: data.structures/java/src/demos/queue/Josephus.java
/**
 * @author Pablo López, Data Structures, Grado en Informática. UMA.
 *
 * Simulation of Flavius Josephus' problem using a Queue.
 */

package demos.queue;

import dataStructures.queue.*;
import java.util.LinkedList;

public class Josephus {
	public static void main(String[] args) {

		final int NUM_PRISONERS = Integer.parseInt(args[0]);
		final int PRISONERS_TO_SKIP = Integer.parseInt(args[1]);

		Queue<Integer> prisoners = new ArrayQueue<>();
		for (int i = 0; i < NUM_PRISONERS; i++) {
			prisoners.enqueue(i);
		}

		LinkedList<Integer> killed = new LinkedList<>();
		while (killed.size() != NUM_PRISONERS - 1) {
			// skip prisoners
			for (int i = 0; i < PRISONERS_TO_SKIP; i++) {
				Integer p = prisoners.first();
				prisoners.dequeue();
				prisoners.enqueue(p);
			}
			// kill the first one
			Integer toBeKilled = prisoners.first();
			prisoners.dequeue();
			killed.addLast(toBeKilled);
		}

		System.out.println("prisoners killed = " + killed);
		System.out.println("prisoner saved = " + prisoners.first());
	}
}
